Architecture.md

```diff
@@ -0,0 +1,287 @@
+roadmap, where each stage builds on the last:
+
+1. Controlling validator set with a privileged contract
+1. Implement PoA via modified multisig contracts (no rewards)
+1. Bootstrapping
+1. Implement PoS via token-locking contract
+1. Distribute rewards to validators (lazy calculated, per epoch, not every block)
+1. Manage Slashing feedback when evidence is submitted, slashing done in contract
+1. PoE as mixer between PoA and PoS implementations
+1. Governance control and voting power
+1. Support dPoS and delegator rewards/slashing
+1. Support dPoE and delegator rewards/slashing
+
+## Contract-Controlled Validators
+
+This involves a separate Go module `x/validator` that gives super powers to one specified CosmWasm
+contract. We can use Custom Message and Queries between CosmWasm and the chain to enable this.
+
+The `x/validator` module state is more or less:
+
+```go
+type Permissions struct {
+    // This account can set any of the addresses in this state
+    Admin       sdk.AccAddress
+    // Address of a CosmWasm contract (CW7) that can set the validator power
+    Authority   sdk.AccAddress
+    // Address of a CosmWasm contract (CW8) that receives rewards and can distribute them
+    // Added in "Distributing Rewards" (if nil, no rewards distributed)
+    Distribution sdk.AccAddress
+    // Address of a CosmWasm contract (CW9) that receives slashing evidence and can punish validators
+    // Added in "Slashing Feedback" (if nil, no slashing occurs)
+    Slashing sdk.AccAddress
+}
+
+type Validator struct {
+    TendermintPubKey    ed25519.PubKey
+    Weight              uint64
+}
+```
+
+The Admin address can set a new contract address or set a new admin address (or clear admin). This is mainly used for bootstrapping, but the admin can remain with some governance vote, to allow swapping out the staking mechanism later.
+
+When the Contract address is changed, the `x/validator` module will make a query to that Contract to get the current
+validator set `[]Validator`. Afterwards, it will process `MsgValidatorDiff` to update/patch the validator set, which must
+be authorized by `Authority` address to be accepted. We will define a CW-7 spec to define the interface of the contract that
+can interact with the `x/validator` module.
+
+`x/validator` implements `EndBlocker` and will send Tendermint any diffs to the validator set that occurred in this block.
+
+## Simple PoA Contract
+
+We have built out the [CW3 multisig spec](https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw3/README.md)
+and are working to [separate groups from the voting conditions](https://github.com/CosmWasm/cosmwasm-plus/issues/80).
+Groups will be defined in CW4.
+
+We can start with a standard, flexible multisig design. The group that stores the voters will implement not only
+CW4, but also CW7 and can be registered on `x/validator`. (Note that CW7 requires the Validator pubkey in addition
```

Comment:
Because we need to set that in `x/validator`, as that is how it can update tendermint consensus.
This is the tendermint ed25519 pubkey, not the sdk pubkey.

I can make it clearer